### ISSUE TITLE
I expect no-unused-vars rule to be skipped if there is an object spread

### ISSUE THREAD

**Issue by gre** [eslint/eslint#4103]
```js
const { foo, ...rest } = obj;
```

**"foo" is defined but never used**
happens,
but in most case you still want to do this because you want to take obj without some named fields.

I personally would expect the no-unused-vars rule to be skipped if there is an object spread.

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by nzakas** [comment]
I'm not sure I can agree with that assertion. It might be best for you to disable the rule for that line when doing so.

**Comment by gre** [comment]
Yeah this issue have been hurting me for long, but it turns out that when I use object spread, I always want this enabled.
Is there a way we could make that a option of the rule?

**Comment by nzakas** [comment]
As object spread is an experimental feature, we aren't going to make special affordances for it until it's finalized. At this time, I don't think we can make any change for this use case. I'd suggest disabling the rule for that line or doing something like this:

```
var const { ...rest } = obj;
delete rest.foo;
```

**Comment by fredguest** [comment]
@nzakas, i agree with @gre that this is not being handled correctly at the moment. for the linter to remain silent when `foo` in the code example above is unused would not be a special affordance, it would be the correct behavior in every instance. the analog is the linter's behavior in this scenario:
```
someCallback(arg1, arg2) {
  return arg2;
}
```
the linter does not warn about arg1 being unused, because it accounts for the fact that it may only be present to access arg2. the scenario with a rest spread is exactly the same; because the destructured variables preceding the rest object may only be present to access the rest object, the linter should never complain if they are unused.

**Comment by BYK** [comment]
I think the best way to handle this is to use [varsIgnorePattern](http://eslint.org/docs/rules/no-unused-vars#ignore-identifiers-that-match-specific-patterns) option and use names that match that pattern.

**Comment by gre** [comment]
having to change the variable names to make this silent does not sound "the best way" to me.
Especially that you can't really change that variable names as they are just the object property names (in my initial example).

Now I understand that destructing assignment with spread object is an experimental feature, not even in ES6, but I think when this gets in and when eslint supports this, we should definitely have an option to ignore the variables you extract out (I don't mind if it's an option but I would personally make it silent by default as @fredguest said. *I think it's one of the only remaining issue I have with eslint when writing ES2015+ code & using with babel-eslint*).

**Comment by BYK** [comment]
>Especially that you can't really change that variable names as they are just the object property names (in my initial example).

Oh, right. I was thinking about array spread for some reason. I agree having an option similar to the one we have for later arguments in functions would be useful when this feature solidifies. Until then, disabling the line looks like the best way forward. Personally, I prefer this over the linter silently ignoring this to make it explicit that my intention was indeed to ignore those.

**Comment by gre** [comment]
fair point

**Comment by slyngbaek** [comment]
@gre I too had this issue. You can use the `varsIgnorePattern` to achieve what you want.

[This](http://eslint.org/docs/rules/no-unused-vars#varsignorepattern) option lets you specify a regex for which to ignore the rule.

```
const {
    foo:_IGNORE_VAR_,
    ...args
} = obj
```

**Comment by smeijer** [comment]
@AsaAyers how's that different from:

```js
// eslint-disable-next-line no-unused-vars
const { foo, bar, baz, ...rest } = obj;
```

That way you can easily delete the comment once the problem has been fixed.

**Comment by AsaAyers** [comment]
It's not significantly better or worse, just another option. I like it because it's not tool-specific and I don't have to remember the right eslint comment syntax. I forgot `eslint-disable-next-line` is a thing, so when I need inline configs I usually surround my code with `eslint-disable <whatever>` and `eslint-enable`.

**Comment by victormiguez** [comment]
If anyone need this just like me today, it`s already possible to do with [no-unused-vars: ["error", { "ignoreRestSiblings": true }]](http://eslint.org/docs/rules/no-unused-vars#ignorerestsiblings).

**Comment by platinumazure** [comment]
@eslint/eslint-team Is this issue fixed by the `"ignoreRestSiblings"` option now present in `no-unused-vars`? (See https://github.com/eslint/eslint/issues/4103#issuecomment-308184779)
